tests/bsim/bluetooth/host/misc/sample_test/src/dut.c

```diff
@@ -0,0 +1,167 @@
+
+static DEFINE_FLAG(is_subscribed);
+
+static void ccc_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	/* assume we only get it for the `test_gatt_service` */
+	if (value) {
+		SET_FLAG(is_subscribed);
+	} else {
+		UNSET_FLAG(is_subscribed);
+	}
+}
+
+BT_GATT_SERVICE_DEFINE(test_gatt_service, BT_GATT_PRIMARY_SERVICE(test_service_uuid),
+		       BT_GATT_CHARACTERISTIC(test_characteristic_uuid,
+					      (BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE |
+					       BT_GATT_CHRC_NOTIFY | BT_GATT_CHRC_INDICATE),
+					      BT_GATT_PERM_READ | BT_GATT_PERM_WRITE, NULL, NULL,
+					      NULL),
+		       BT_GATT_CCC(ccc_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),);
+
+static void setup(char *message)
+{
+	/* Initialize the bsim framework, and trigger a callback before the
+	 * simulation times out.
+	 */
+	TEST_START(timeout_get(), message);
+
+	bs_sync_init();
+
+	int err = bt_enable(NULL);
+
+	TEST_ASSERT(err == 0, "Can't enable Bluetooth (err %d)\n", err);
+
+	LOG_DBG("Bluetooth initialized.");
+}
+
+/* This is the entrypoint for the DUT.
+ *
+ * This is executed by the `bst_test` framework that comes with the bsim
+ * distribution. The framework selects which "main" function to run as
+ * entrypoint depending on the `-testid=` command-line parameter passed to the
+ * zephyr executable.
+ *
+ * In our case, the `testid` is set to "dut" and `entrypoint_dut()` is mapped to
+ * the "dut" ID in `entrypoints[]`.
+ *
+ * In our case we only have two entrypoints, as we only have a single test
+ * involving two devices (so 1x 2 entrypoints). One can define more test cases
+ * with different entrypoints and map them to different test ID strings in
+ * `entrypoints[]`
+ */
+void entrypoint_dut(void)
+{
+	/* Please leave a comment indicating what the test is supposed to test,
+	 * and what is the pass verdict. A nice place is at the beginning of
+	 * each test entry point. Something like the following:
+	 */
+
+	/* Test purpose:
+	 *
+	 * Verifies that we are able to send a notification to the peer when
+	 * `CONFIG_BT_GATT_ENFORCE_SUBSCRIPTION` is disabled and the peer has
+	 * unsubscribed from the characteristic in question.
+	 *
+	 * Two devices:
+	 * - `dut`: tries to send the notification
+	 * - `peer`: will receive the notification
+	 *
+	 * Procedure:
+	 * - [dut] establish connection to `peer`
+	 * - [peer] discover GATT and subscribe to the test characteristic
+	 * - [dut] send notification #1
+	 * - [peer] wait for notification
+	 * - [peer] unsubscribe
+	 * - [dut] send notification #2
+	 * - [peer] and [dut] pass test
+	 *
+	 * [verdict]
+	 * - peer receives notifications #1 and #2
+	 */
+	int err;
+	bt_addr_le_t peer = {};
+	struct bt_conn *conn = NULL;
+	const struct bt_gatt_attr *attr;
+	uint8_t data[10];
+
+	/* Initialize framework and enable Bluetooth */
+	setup("dut");
+
+	/* Find the address of the peer */
+	err = bt_testlib_scan_find_name(&peer, CONFIG_BT_DEVICE_NAME);
```

Comment:
This builds on the assumption that both devices have the same name, right? 

It's OK, but I think assumptions should be written as explicit comments